spring-cloud-skipper-server-core/src/main/java/org/springframework/cloud/skipper/server/deployer/strategies/HealthCheckAndDeleteStep.java

```diff
@@ -0,0 +1,153 @@
+
+	private final Logger logger = LoggerFactory.getLogger(SimpleRedBlackUpgradeStrategy.class);
+
+	private final ReleaseRepository releaseRepository;
+
+	private final DeployerRepository deployerRepository;
+
+	private final DeleteStep deleteStep;
+
+	private final HealthCheckProperties healthCheckProperties;
+
+	public HealthCheckAndDeleteStep(ReleaseRepository releaseRepository,
+			DeployerRepository deployerRepository,
+			DeleteStep deleteStep,
+			HealthCheckProperties healthCheckProperties) {
+		this.releaseRepository = releaseRepository;
+		this.deployerRepository = deployerRepository;
+		this.deleteStep = deleteStep;
+		this.healthCheckProperties = healthCheckProperties;
+	}
+
+	@Async("skipperThreadPoolTaskExecutor")
+	@Transactional
+	public CompletableFuture<Void> waitForNewAppsToDeploy(Release existingRelease,
+			AppDeployerData existingAppDeployerData,
+			List<String> applicationNamesToUpgrade, Release replacingRelease,
+			AppDeployerData replacingAppDeployerData) {
+		try {
+			logger.info("Waiting for apps in release {}-v{} to be healthy.", replacingRelease.getName(),
+					replacingRelease.getVersion());
+			boolean healthy = this.isHealthy(replacingRelease, replacingAppDeployerData);
+
+			if (healthy) {
+				logger.info("Apps in release {}-v{} are healthy.", replacingRelease.getName(),
+						replacingRelease.getVersion());
+				logger.info("Deleting changed applications from previous release {}-v{}",
+						existingRelease.getName(),
+						existingRelease.getVersion());
+
+				this.deleteStep.delete(existingRelease, existingAppDeployerData, applicationNamesToUpgrade);
+
+				// Update Status in DB
+				Status status = new Status();
+				status.setStatusCode(StatusCode.DEPLOYED);
+				replacingRelease.getInfo().setStatus(status);
+				replacingRelease.getInfo().setDescription("Upgrade complete");
+				this.releaseRepository.save(replacingRelease);
+				logger.info("Release {}-v{} has been DEPLOYED", replacingRelease.getName(),
+						replacingRelease.getVersion());
+			}
+			else {
+				logger.error("New release " + replacingRelease.getName() + " was not detected as healthy." +
+						"Keeping existing releases.");
+				Status status = new Status();
+				status.setStatusCode(StatusCode.DEPLOYED);
+				replacingRelease.getInfo().setStatus(status);
+				replacingRelease.getInfo().setStatus(status);
+				replacingRelease.getInfo().setDescription("Did not detect apps in release as healthy.");
+				this.releaseRepository.save(replacingRelease);
+			}
+		}
+		catch (Exception e) {
+			logger.error("Error upgrading new app", e);
+			// Update Status in DB
+			Status status = new Status();
+			status.setStatusCode(StatusCode.FAILED);
+			replacingRelease.getInfo().setStatus(status);
+			replacingRelease.getInfo().setDescription("Failed upgrade. Exception = [" +
+					e.getClass().getName() + "], Message = [" + e.getMessage() + "]");
+			this.releaseRepository.save(replacingRelease);
+		}
+		return null;
```

Comment:
U need to actually return `CompletableFuture` which is then completed with `complete(T)` or `completeExceptionally(Throwable)` instead of returning `null`.

Reply:
I think actual stuff what this method do should be fully wrapped in `Future` which is then executed when future value is requested.

Reply:
I'll change this to void, it ended up being a `CompletableFuture` because I made some mistakes not having the method in a separate class and was spinning my wheels for a bit.